src/eg/src/guardian_coeff_proof.rs

```diff
@@ -0,0 +1,223 @@
+        // v = 0x10 | b(i,4) | b(j,4) | b(coefficient,512) | b(h,512)
+        let mut v = vec![0x10];
+        v.extend_from_slice(i.to_be_bytes().as_slice());
+        v.extend_from_slice(j.to_be_bytes().as_slice());
+        v.extend_from_slice(to_be_bytes_left_pad(coefficient, 512).as_slice());
+        v.extend_from_slice(to_be_bytes_left_pad(h, 512).as_slice());
+        let c = eg_h(&h_p, &v);
+        //The challenge is not reduced modulo q (cf. Section 5.4)
+        BigUint::from_bytes_be(c.0.as_slice())
+    }
+
+    /// This function computes a single coefficient NIZK proof as specified in Section 3.2.2
+    /// The arguments are
+    /// - csprng - secure randomness generator
+    /// - fixed_parameters - the fixed parameters
+    /// - h_p - the parameter base hash
+    /// - i - the guardian index
+    /// - j - the coefficient index
+    /// - coefficient - the guardian's secret coefficient
+    /// - commitment - the coefficient commitment
+    pub fn new(
+        csprng: &mut Csprng,
+        fixed_parameters: &FixedParameters,
+        h_p: HValue,
+        i: u32,
+        j: u32,
+        coefficient: &SecretCoefficient,
+        commitment: &CoefficientCommitment,
+    ) -> Self {
+        let coefficient = &coefficient.0;
+        let commitment = &commitment.0;
+        // Compute commit message
+        let u = csprng.next_biguint_lt(fixed_parameters.q.as_ref());
+        let h = fixed_parameters.g.modpow(&u, fixed_parameters.p.as_ref());
+        // Compute challenge
+        let c = Self::challenge(h_p, i, j, commitment, &h);
+        // Compute response
+        let s = &c * coefficient;
+        let v = fixed_parameters.q.subtract_group_elem(&u, &s);
+        CoefficientProof {
+            challenge: c,
+            response: v,
+        }
+    }
+
+    /// This function verifies a [`CoefficientProof`] with respect to a given [`CoefficientCommitment`] and context
+    /// The arguments are
+    /// - self - the `CoefficientProof`
+    /// - fixed_parameters - the fixed parameters
+    /// - h_p - the parameter base hash
+    /// - i - the guardian index
+    /// - j - the coefficient index
+    /// - commitment - the coefficient commitment
+    pub fn validate(
+        &self,
+        fixed_parameters: &FixedParameters,
+        h_p: HValue,
+        i: u32,
+        j: u32,
+        commitment: &CoefficientCommitment,
+    ) -> Result<(), ProofValidationError> {
+        let commitment = &commitment.0;
+        let zero = BigUint::from(0u8);
+        let one = BigUint::from(1u8);
+        let p = fixed_parameters.p.as_ref();
+        let q = fixed_parameters.q.as_ref();
+        let g = &fixed_parameters.g;
+
+        // Verification check (2.A)
+        if &zero > commitment || commitment >= p || commitment.modpow(q, p) != one {
+            return Err(ProofValidationError::CommitmentNotInGroup);
+        }
+        // Verification check (2.B)
+        if zero > self.response || &self.response >= q {
```

Comment:
We could remove the `0 <= ` checks as `BigUint` always satisfy this.

Reply:
Yes, agree.